File: api/server/app/de/htwg/zeta/server/generator/generators/diagram/LinkhelperGenerator.scala
package de.htwg.zeta.server.generator.generators.diagram

import de.htwg.zeta.server.generator.model.diagram.Diagram
import de.htwg.zeta.server.generator.model.diagram.edge.Edge

import scala.util.Success
import scala.util.Try

/**
 * Created by julian on 10.02.16.
 */
object LinkhelperGenerator {
  def generate(diagram: Diagram) = {
    s"""
    $generateHead
    ${generateLinkhelper(diagram)}
    """
  }

  def generateHead =
    """
    /*
     * This is a generated linkhelper file for JointJS
     */
    """

  protected def generateLinkhelper(diagram: Diagram) =
    s"""
    var linkhelper = {
      ${generatePlacingTexts(diagram)}
      $generateHelperMethods
      ${generateMapping(diagram.edges)}
    };
    """

  protected def generatePlacingTexts(diagram: Diagram) = {
    s"""
    placingTexts:{
      ${diagram.edges.map(e => s"""${e.name}: { ${generateStringProperties(e).mkString(",")} }""").mkString(",")}
    },
    """
  }

  protected def generateStringProperties(edge: Edge) = {
    val stringProperties = edge.connection.vals
    for {(key, value) <- stringProperties} yield s"""'$key': "${value.getOrElse("").toString}""""
  }

  protected def generateHelperMethods =
    """
    getLabelText: function(edge, textId){
      var text = this.placingTexts[edge][textId];
      if(text === undefined){
        text = "";
      }
      return text;
    },
    """

  private def generateMapping(edges: Iterable[Edge]) =
    s"""
      mapping: {
        ${edges.map(generateMappingEntry).mkString(",")}
      }
    """

  private def generateMappingEntry(edge: Edge) =
  s"""
    ${edge.mcoreElement.name}: {
      ${edge.mcoreElement.attributes.map(attribute => generateMappingEntryAttribute(edge, attribute.name)).mkString(",")}
    }
  """

  private def generateMappingEntryAttribute(edge: Edge, name: String): String = {
    Try(edge.connection.vars.filter{case (attribute, _) => attribute.name == name}.head) match {
      case Success(i) => i._2 match {
        case Some(text) => s"""${text.id}: "$name" """
        case _ => ""
      }
      case _ => ""
    }
  }
}

File: webapp/src/graphical-editor/model/generator/diagram/LinkHelperGenerator.js
function findEdge(diagram, name) {
    if (diagram && diagram.model && diagram.model.edges) {
        return diagram.model.edges.find(e => e.name === name);
    }
    return null;
}

function getEdgeVal(edge, key) {
    const val = findVal(edge, key);
    return val ? val.value : '';
}

function findVal(edge, key) {
    if (edge && edge.connection && edge.connection.vals) {
        return edge.connection.vals.find(v => v.key === key);
    }
    return null;
}

function createMappingEdges(diagram) {
    const mapper = {};
    diagram.model.edges.forEach(edge => {
        const attributes = edge.connection && edge.connection.vars ? createMappingAttribute(edge) : {};
        mapper[edge.mReference] = attributes;
    });
    return mapper;
}

function createMappingAttribute(edge) {
    const mapper = {};
    edge.connection.vars.forEach(entry => {
        mapper[entry.key] = entry.value;
    });
    return mapper;
}

/**
 * Generator of linkhelper for JointJS to map Reference attribute to
 * textbox and resolve label text
 */
export default class LinkHelperGenerator {

    constructor(diagram) {
        this.diagram = diagram;
    }

    getLabelText(edgeName, textId) {
        const edge = findEdge(this.diagram, edgeName);
        return getEdgeVal(edge, textId);
    }

    get mapping() {
        const condition = this.diagram && this.diagram.model && this.diagram.model.edges;
        return condition ? createMappingEdges(this.diagram) : {};
    }
}